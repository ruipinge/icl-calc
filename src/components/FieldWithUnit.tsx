import * as React from 'react';

import { Field, ErrorMessage } from 'formik';
import { getClassName } from '../util';

type Unit = 'mm' | 'nm' | 'º' | 'dpt' | 'μm';

const UNITS: Map<Unit, string> = new Map([
  ['mm', 'millimetres'],
  ['nm', 'nanometres'],
  ['º', 'degrees'],
  ['dpt', 'dioptres'],
  ['μm', 'micrometres']
]);

interface FieldWithUnitProps {
  label: string;
  name: string;
  unit: Unit;
  touched?: boolean;
  placeholder?: string;
  error?: string;
  value?: number;
  disabled?: boolean;
}

const UnitSufix = ({ unit }: { unit: Unit }) => (
  <div className="input-group-append">
    <span className="input-group-text" title={UNITS.get(unit)}>
      {unit}
    </span>
  </div>
);

export const FieldWithUnit: React.FC<FieldWithUnitProps> = ({
  label,
  name,
  unit,
  placeholder,
  error,
  touched,
  value,
  disabled
}) => {
  return (
    <div className={'form-group row'}>
      <label htmlFor={name + 'field'} className="col-sm-6 col-form-label">
        {label}
      </label>
      <div className="col-sm-6">
        <div className="input-group">
          {disabled ? (
            <input
              name={name}
              value={value}
              className="form-control text-right"
              disabled={true}
            />
          ) : (
            <Field
              type="number"
              name={name}
              className={getClassName({
                error: error,
                touched: touched,
                base: ['form-control', 'text-right']
              })}
              placeholder={placeholder}
              autoComplete="off"
            />
          )}
          <UnitSufix unit={unit} />
          <ErrorMessage
            name={name}
            component="div"
            className="invalid-feedback"
          />
        </div>
      </div>
    </div>
  );
};
