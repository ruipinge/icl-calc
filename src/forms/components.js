import React from 'react';

import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';
import InputGroup from 'react-bootstrap/InputGroup';
import Row from 'react-bootstrap/Row';


export const UnitInput = class UnitInput extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: props.defaultValue
    };

    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(event) {
    this.setState({
      value: event.target.value
    });
  }

  render() {
    const label = this.props.label,
      unit = this.props.unit,
      placeholder = this.props.placeholder,
      textAlign = this.props.placeholder || 'right';
    return (
      <Form.Group as={Row} controlId={this.props}>
        <Form.Label column sm="7">{label}</Form.Label>
        <Col sm="5">
          <InputGroup>
            <Form.Control type="text" className={'text-' + textAlign} value={this.state.value} onChange={this.handleChange} placeholder={placeholder} autoComplete="off" />
            {unit
              ? <InputGroup.Append>
                  <InputGroup.Text>{unit}</InputGroup.Text>
                </InputGroup.Append>
              : null
            }
          </InputGroup>
        </Col>
      </Form.Group>
    );
  };
};
