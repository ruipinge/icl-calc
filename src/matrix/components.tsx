import { LENS_SIZES, MYOPIA_LEVELS } from './data';

export const MatrixRow = ({
  label,
  title,
  values
}: {
  label: string;
  title: string;
  values: number[];
}) => (
  <tr>
    <th scope="row" title={title}>
      {label}
    </th>
    {values.map((val, i) => (
      <td key={i} className="text-right">
        {val}
      </td>
    ))}
  </tr>
);

export const DividerRow = () => (
  <tr className="table-bordered">
    <td colSpan={LENS_SIZES.length * MYOPIA_LEVELS.length + 1} />
  </tr>
);
