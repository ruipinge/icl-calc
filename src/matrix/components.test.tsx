import { DividerRow, MatrixRow } from './components';

import { render } from '@testing-library/react';

it('renders DividerRow without crashing', () => {
  const { asFragment } = render(
    <table>
      <tbody>
        <DividerRow />
      </tbody>
    </table>
  );
  expect(asFragment()).toMatchSnapshot();
});

it('renders MatrixRow without crashing', () => {
  const { asFragment } = render(
    <table>
      <tbody>
        <MatrixRow label="Some label" title="Some title" values={[1, 2, 3]} />
      </tbody>
    </table>
  );
  expect(asFragment()).toMatchSnapshot();
});
