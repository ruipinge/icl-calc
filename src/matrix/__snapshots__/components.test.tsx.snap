// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DividerRow without crashing 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr
        class="table-bordered"
      >
        <td
          colspan="10"
        />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`renders MatrixRow without crashing 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <th
          scope="row"
          title="Some title"
        >
          Some label
        </th>
        <td
          class="text-right"
        >
          1
        </td>
        <td
          class="text-right"
        >
          2
        </td>
        <td
          class="text-right"
        >
          3
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
